### ISSUE TITLE
User-made shader functions don't work if used before being defined

### ISSUE THREAD

**Issue by LikeLakers2** [godotengine/godot#17905]
**Godot version:**
Godot Engine v3.0.2.stable.official

**OS/device including version:**
Windows 10 Home v1709 (Fall Creator's Update); Nvidia GTX 660

**Issue description:**
User-made shader functions just simply do not work. The editor recognizes functions for the purposes of auto-completion, but it says it can't find the functions when you attempt to use them.

**Steps to reproduce:**
1. Create a shader that utilizes a user-made function (not a built-in one).
2. Godot will complain about being unable to find the shader function, despite also being able to auto-complete to the new function. Additionally, the shader will not work in-game.

**Minimal reproduction project:**
[Shader Functions Not Working.zip](https://github.com/godotengine/godot/files/1911710/Shader.Functions.Not.Working.zip)  (shader is within shader.tres if you need to quickly get to it)

-----

Edit: As discussed below, this appears to be an issue with documentation, not exactly a bug, as the GLSL spec says shader functions need to be defined before their first use. This isn't exactly made clear without visiting the GLSL specs, so it'd be a good idea to document it.

**Comment by Zylann** [comment]
It works if you put the function before its usage in the shader:

```glsl
shader_type canvas_item;

vec4 get_color() {
	return vec4(0.5, 0.5, 0.5, 1.0);
}

void fragment() {
	//COLOR = vec4(0.5, 0.5, 0.5, 1.0);
	COLOR = get_color();
}
```

So we need to fix function declaration lookup.

**Comment by LikeLakers2** [comment]
So just to make sure I'm understanding, you're saying you think this is something with GLSL in general? Huh. Will this be fixed, or is this more a documentation issue than anything?

**Comment by mrcdk** [comment]
Yes, it's part of the spec. A function must be declared or defined before being called. 

![https://i.imgur.com/gzAbJhX.png](https://i.imgur.com/gzAbJhX.png)

From here: https://www.khronos.org/registry/OpenGL/specs/gl/GLSLangSpec.4.40.pdf (6.1 Function Definitions)

**Comment by Zylann** [comment]
Is it possible to declare prototypes in the Godot shading language though?

**Comment by mrcdk** [comment]
Doesn't look like it or overloading a function. 

I was answering the question about if this is something specific to godot or GLSL.

**Comment by LikeLakers2** [comment]
Alright, well it'd be nice if this was documented. Or, more helpfully, if order wasn't an issue. Though I'm unsure if Godot would be willing to deviate from the spec a little for the sake of making it easier to do a shader.

**Comment by vnen** [comment]
The problem is if the spec doesn't accept it, then Godot would have to reorder everything. Or declare a prototype for all functions in the beginning.

**Comment by LikeLakers2** [comment]
Perhaps to avoid confusion, Godot could also supply a more descriptive error, if possible? If someone tries to use a function before it's defined, then Godot could notify the user that shader functions need to be before their usage. This just seems like the sort of thing that would trip up other people trying to learn shaders.

**Comment by willvincent** [comment]
Deviating from the spec would be a bad idea. Docs should probably reference the spec, doesn't seem like it's explicitly necessary to call out _just this_ specific part of spec in the docs though to me.

**Comment by LikeLakers2** [comment]
I'm not sure how supplying a more descriptive error would be deviating from the spec... or were you talking about making it possible to define them after?

**Comment by willvincent** [comment]
Defining user functions after their usage is deviation from spec, that shouldn't happen.  As for error message, maybe it makes sense to be more clear, but not defining a function after it's usage should be kind of common sense really -- and as that is part of the spec, the docs probably should just refer to the spec, rather than reiterating and spelling things out. And to simply spell out this single part of the spec in the docs, but not other parts of the spec seems silly.

I guess my position on it is, the spec is not Godot's, so docs should refer people to the actual spec and indicate that it be followed, rather than trying to explicitly spell out every detail of it.

**Comment by LikeLakers2** [comment]
In regards to the suggestion to use prototypes (if it's even possible): Would that be deviating from spec? Like, I can see why it might be considered deviating from spec since we're allowing it, but I'm not sure it's *actually* deviating from spec since we're already processing the code and turning it into proper GLSL code, which we then compile. Because if not, I would be all for that suggestion, as it makes the shading language more consistent in comparison to GDScript.

----

@willvincent 
Honestly, I'm confused. Why would using a function only after it's been defined be common sense? I'm not going to argue that we shouldn't follow the spec, mind you, just... I'm confused about the "it's common sense" bit. It's not really common sense in my eyes, because I've seen many languages that don't seem to care where functions are defined so long as they're defined. Even GDScript lets you use functions defined after their usage. Again, this isn't saying we have to deviate from the spec. I'm just confused about your "it's common sense" bit.

**Comment by willvincent** [comment]
@LikeLakers2 I dunno.. it's common sense to me. lol

**Comment by akien-mga** [comment]
Fixed by godotengine/godot-docs#2046.
